### DISCUSSION TITLE
Retrieving instrument data with multiple parameters & lambda expressions?

### DISCUSSION THREAD

**Discussion by Gazelle79** [microsoft/Qcodes#5393] [Q&A]
Hi, all.

I'm using QCodes to talk to an API that uses HTTP GET and HTTP POST to get and retrieve data, respectively. My code inherits from a QCodes instrument class. All of this is in Python.

The method get_device_info(device_name, device_id, active) requires three parameters in order to return data related to it. Assume it lives elsewhere in this same python class. The python class inherits from QCodes.Instrument, so add_parameter() is available. I connect my method to the device by initializing the following code in the constructor, like this:

```
    self.add_parameter(name = "this_device",
                                    label = "Example Device",
                                    get_cmd = lambda x: self.get_device_info( x["device_name"], x["device_id"], x["active""]             
                                   ) 
```
...but I keep getting an error message:

> **Exception has occured: KeyError**
> 'Duplicate parameter name channel on instrument <Controller: This_Device>'

There are no duplicate keys anywhere that I can see.

Again, all three parameters are required, and I'm using them to read information about a device. Error messages like this tell me that something went wrong behind the scenes. QCodes was expecting things in a certain format, and it didn't receive them in the way it was expecting.

How can I use a Lambda expression to account for parameters that must be passed in, so I can retrieve data? Should I use something else? 

Please let me know.

**Comment by jenshnielsen** [comment]
There is no requirement that the function that you pass to get_cmd is a lamda so I would suggest making in a regular private method if you find the lambda too limiting. 

But I actually think the problem is elsewhere. It looks like you have created a parameter called channel more than once on the same instrument.
